<a href="https://colab.research.google.com/github/kevinokelly23-collab/Data-Acquisition-Project/blob/main/%20notebooks/Netflix_Data_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Acquisition Project
# Kevin O'Kelly - Netflix Movies and Box office analysis


Project overview


This project follows the Data Acquisition coursework brief, which requires:
1. Importing and exploring a Kaggle dataset.  
2. Web scraping a complementary dataset.  
3. Merging the datasets.  
4. Visualising insights with Seaborn.  
5. Documenting and reflecting on AI use.


Datasets

file path: notebooks/Netflix_Data_Project.ipynb

Primary - Kaggle: Netflix Movies and TV Shows  https://www.kaggle.com/datasets/shivamb/netflix-shows

Complimentary - Scraped: Box Office Mojo All Time Worldwide Box Office  
https://www.the-numbers.com/box-office-records/worldwide/all-movies/cumulative/all-time/2001

# Step 1 Importing, exploring and cleaning my Kaggle Dataset

In [ ]:
#Step 1: Import Libraries and Load netflix dataset
import pandas as pd
# Load the dataset
df = pd.read_csv("/content/netflix_titles.csv")


In [35]:
#Step 1.1: Exploring the dataset
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [ ]:
df.describe(include='all')

show_id   type   title       director                cast  \
count     8807   8807    8807           6173                7982   
unique    8807      2    8807           4528                7692   
top      s8807  Movie  Zubaan  Rajiv Chilaka  David Attenborough   
freq         1   6131       1             19                  19   
mean       NaN    NaN     NaN            NaN                 NaN   
std        NaN    NaN     NaN            NaN                 NaN   
min        NaN    NaN     NaN            NaN                 NaN   
25%        NaN    NaN     NaN            NaN                 NaN   
50%        NaN    NaN     NaN            NaN                 NaN   
75%        NaN    NaN     NaN            NaN                 NaN   
max        NaN    NaN     NaN            NaN                 NaN   

              country       date_added  release_year rating  duration  \
count            7976             8797   8807.000000   8803      8804   
unique            748             1767           NaN     17       220   
top     United States  January 1, 2020           NaN  TV-MA  1 Season   
freq             2818              109           NaN   3207      1793   
mean              NaN              NaN   2014.180198    NaN       NaN   
std               NaN              NaN      8.819312    NaN       NaN   
min               NaN              NaN   1925.000000    NaN       NaN   
25%               NaN              NaN   2013.000000    NaN       NaN   
50%               NaN              NaN   2017.000000    NaN       NaN   
75%               NaN              NaN   2019.000000    NaN       NaN   
max               NaN              NaN   2021.000000    NaN       NaN   

                           listed_in  \
count                           8807   
unique                           514   
top     Dramas, International Movies   
freq                             362   
mean                             NaN   
std                              NaN   
min                              NaN   
25%                              NaN   
50%                              NaN   
75%                              NaN   
max                              NaN   

                                              description  
count                                                8807  
unique                                               8775  
top     Paranormal activity at a lush, abandoned prope...  
freq                                                    4  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [ ]:
df['type'].value_counts()


type
Movie      6131
TV Show    2676
Name: count, dtype: int64

In [ ]:
df['release_year'].describe()


count    8807.000000
mean     2014.180198
std         8.819312
min      1925.000000
25%      2013.000000
50%      2017.000000
75%      2019.000000
max      2021.000000
Name: release_year, dtype: float64

Okay so now that I have gotten my kaggle dataset imported and I can see that eveyrthing is as should be I want to filter to movies only as I will not be using the TV shows.

In [ ]:
# Step 1.3: Filtering to movies only as I will not be using the TV shows

movies_df = df[df['type'].str.strip().str.lower() == 'movie'].copy()



In [ ]:
movies_df['type'].value_counts()


type
Movie    6131
Name: count, dtype: int64

The Kaggle dataset 'Netflix Movies and TV shows' contains 8807 titles available.

After filtering to "movie" the datset contains 6131 movies.

Most movies were released between 2013-2021.

In [ ]:
#Step 1.4: Checking for missing values

movies_df.isna().sum().sort_values(ascending=False)


cast            475
country         440
director        188
duration          3
rating            2
title             0
show_id           0
type              0
release_year      0
date_added        0
listed_in         0
description       0
dtype: int64

# Step 1.41 Observations
**Data quality and Missing Values**

The dataset is generally clean, while only minor missing values:

Director : 188 missing

cast : 475 missing

country : 440 missing

duration : 3 missing

rating : 2 missing

**High missing Columns**

Director : 188 missing

cast : 475 missing

country : 440 missing



*   These three fields do not determine merges, do not affect numerical analysis and are not used for modelling



*   Fill with unknown

**Low missing columns**

Duration : 3 missing

Rating : 2 missing


Duration is important because its a clear movie attribute however only 3 rows are missing and same with rating


*   drop the rows











In [ ]:
# Step 1.5: Handling Missing Values

# Fill high-missing categorical metadata
movies_df['cast'] = movies_df['cast'].fillna('Unknown')
movies_df['country'] = movies_df['country'].fillna('Unknown')
movies_df['director'] = movies_df['director'].fillna('Unknown')

# Drop rare missing numerical/categorical rows
movies_df = movies_df.dropna(subset=['duration'])
movies_df = movies_df.dropna(subset=['rating'])

In [ ]:
movies_df.isna().sum()


show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

In [38]:
movies_df['rating'].isna().sum()
movies_df['duration'].isna().sum()


np.int64(0)




*   As we can see here we have successfully removed all missing values. 0 in each column.





In [ ]:
movies_df.shape


(6126, 12)



*   I am going to convert the duration column to an integer so it can be used for numerical analysis, as the original “90 min” text format cannot be processed in calculations.




In [37]:
# Changing to integer

movies_df['duration'] = (
    movies_df['duration']
        .str.replace(' min', '', regex=False)
        .astype(int)
)


AttributeError: Can only use .str accessor with string values!

In [ ]:
# Validation
movies_df['duration'].head()
movies_df['duration'].dtype


dtype('int64')

In [ ]:
movies_df['duration'].mean()


np.float64(99.58488410055502)



*   I am going to convert date_added into a proper datetime format so that the column can be sorted, filtered by time, and used for any time-based analysis, since the original text format cannot be processed reliably.




In [ ]:
movies_df['date_added'] = pd.to_datetime(movies_df['date_added'])


In [ ]:
movies_df['date_added'].dtype


dtype('<M8[ns]')

# Step 1 Synopsis

*   As we can see there is no longer any missing values and the 5 rows have been removed.

*   The original shape was 6131 and its down to 6126.



*   Duration has been changed to an integer.
*   date_ added has been changed to a proper datetime format.



*   The dataset is now cleaned and we can save it and begin on our web scraping.



# Step 2 Web Scraping a Complementary Dataset

The goal of this step is to obtain **box-office performance data** for movies, which is not included in the Kaggle Netflix dataset.  
This information will later be merged using the movie **Title** column to explore relationships between Netflix availability and box-office success.

I will scrape data from *The Numbers* website, which lists the all-time worldwide box-office rankings for movies.


#### Inspecting the Website HTML Structure

Before starting to scrape, I inspected the HTML structure of *The Numbers*
box-office pages.

Each page has a clear and consistent layout:
- The main data is inside a `<center>` tag.
- Within it, there is a `<table>` containing the movie data.
- The `<thead>` section holds the column headers (e.g. *Rank, Movie, Worldwide, Domestic, International, Year*).
- The `<tbody>` contains multiple `<tr>` rows, each representing one movie.
- Each `<td>` cell within a row stores the actual values.

I'm hoping the structure makes it straightforward to scrape the data using BeautifulSoup, as the movie table can be selected directly using:
```python
soup.find("center").find("table")


In [ ]:
# Step 2.1: test one page


import requests
from bs4 import BeautifulSoup
import pandas as pd
import time

url = "https://www.the-numbers.com/box-office-records/worldwide/all-movies/cumulative/all-time/1"

headers = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/123.0.0.0 Safari/537.36"
    )
}

response = requests.get(url, headers=headers)
print("Status code:", response.status_code)

soup = BeautifulSoup(response.text, "html.parser")
center_tag = soup.find("center")
print("Center tag found:", bool(center_tag))

if center_tag:
    table = center_tag.find("table")
    print("Table found:", bool(table))
else:
    print("⚠️ No <center> tag found.")



Status code: 200
Center tag found: True
Table found: False


The status code 200 confirmed the request was successful and the site allowed access.

The center tag existed and showing that part of the page structure was correct, but no table element was detected inside

I asked ChatGPT what I should try next. It suggested broadening the search to look for a table both inside the center tag and anywhere else on the page as a fallback leading to the next step.

In [ ]:
# Step 2.11: Locating the Table Correctly

soup = BeautifulSoup(response.text, "html.parser")

# Try multiple methods to locate the table
table = None

# Option 1: look for <center> first, then any table within
center_tag = soup.find("center")
if center_tag:
    table = center_tag.find("table")

# Option 2: fallback – look for any table on the page
if not table:
    table = soup.find("table")

print("Center tag found:", bool(center_tag))
print("Table found:", bool(table))

if table:
    print("✅ Table located successfully.")
else:
    print("⚠️ Still no table — may need to inspect HTML structure again.")


Center tag found: True
Table found: True
✅ Table located successfully.


By inspecting the website’s structure, I found that the movie table is nested inside a section called page_filling_chart, which explained why it wasn’t detected at first. I updated the scraping logic to locate the table more flexibly within the page.

I learned how conditional searching improves resilience when scraping sites with inconsistent layouts.


In [ ]:
# Step 2.2: Extract Headers and Sample Rows


rows = table.find_all("tr")
headers = [th.text.strip() for th in rows[0].find_all("th")]
print("Headers found:", headers)

sample_data = []
for row in rows[1:6]:
    cols = [td.text.strip() for td in row.find_all("td")]
    sample_data.append(cols)

pd.DataFrame(sample_data, columns=headers)

Headers found: ['Rank', 'Year', 'Movie', 'WorldwideBox Office', 'DomesticBox Office', 'InternationalBox Office']


Rank  Year                                 Movie WorldwideBox Office  \
0    1  2009                                Avatar      $2,923,706,026   
1    2  2019                     Avengers: Endgame      $2,717,503,922   
2    3  2022              Avatar: The Way of Water      $2,322,902,023   
3    4  1997                               Titanic      $2,223,048,786   
4    5  2015  Star Wars Ep. VII: The Force Awakens      $2,056,046,835   

  DomesticBox Office InternationalBox Office  
0       $785,221,649          $2,138,484,377  
1       $858,373,000          $1,859,130,922  
2       $688,809,501          $1,634,092,522  
3       $674,460,013          $1,548,588,773  
4       $936,662,225          $1,119,384,610

# Step 2.2: Observation

The headers and rows were successfully extracted, confirming that the table structure contains the expected columns such as Rank, Year, Movie, WorldwideBox Office, DomesticBox Office, InternationalBox Office

That verfifies that the dat is ready to be scraped.

In [ ]:
# Step 2.3: Full Scraper

from urllib.request import Request, urlopen
from bs4 import BeautifulSoup
import pandas as pd
import time

records = []

for page in range(1, 101):
    url = "https://www.the-numbers.com/box-office-records/worldwide/all-movies/cumulative/all-time"
    if page > 1:
        url += f"/{page}"

    # Add browser header to bypass 403
    req = Request(url, headers={"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"})
    html = urlopen(req)
    soup = BeautifulSoup(html, "html.parser")

    table = soup.find("table")
    rows = table.find_all("tr")

    for row in rows[1:]:
        cols = row.find_all("td")
        if len(cols) == 6:
            data = [col.get_text(strip=True) for col in cols]
            records.append(data)

    print(f"✅ Scraped page {page}")
    time.sleep(1.5)

df = pd.DataFrame(records, columns=["Rank", "Year", "Title", "Worldwide", "Domestic", "International"])



✅ Scraped page 1
✅ Scraped page 2
✅ Scraped page 3
✅ Scraped page 4
✅ Scraped page 5
✅ Scraped page 6
✅ Scraped page 7
✅ Scraped page 8
✅ Scraped page 9
✅ Scraped page 10
✅ Scraped page 11
✅ Scraped page 12
✅ Scraped page 13
✅ Scraped page 14
✅ Scraped page 15
✅ Scraped page 16
✅ Scraped page 17
✅ Scraped page 18
✅ Scraped page 19
✅ Scraped page 20
✅ Scraped page 21
✅ Scraped page 22
✅ Scraped page 23
✅ Scraped page 24
✅ Scraped page 25
✅ Scraped page 26
✅ Scraped page 27
✅ Scraped page 28
✅ Scraped page 29
✅ Scraped page 30
✅ Scraped page 31
✅ Scraped page 32
✅ Scraped page 33
✅ Scraped page 34
✅ Scraped page 35
✅ Scraped page 36
✅ Scraped page 37
✅ Scraped page 38
✅ Scraped page 39
✅ Scraped page 40
✅ Scraped page 41
✅ Scraped page 42
✅ Scraped page 43
✅ Scraped page 44
✅ Scraped page 45
✅ Scraped page 46
✅ Scraped page 47
✅ Scraped page 48
✅ Scraped page 49
✅ Scraped page 50
✅ Scraped page 51
✅ Scraped page 52
✅ Scraped page 53
✅ Scraped page 54
✅ Scraped page 55
✅ Scraped page 56
✅

In [ ]:
df.head()

Rank  Year                                 Title       Worldwide  \
0    1  2009                                Avatar  $2,923,706,026   
1    2  2019                     Avengers: Endgame  $2,717,503,922   
2    3  2022              Avatar: The Way of Water  $2,322,902,023   
3    4  1997                               Titanic  $2,223,048,786   
4    5  2015  Star Wars Ep. VII: The Force Awakens  $2,056,046,835   

       Domestic   International  
0  $785,221,649  $2,138,484,377  
1  $858,373,000  $1,859,130,922  
2  $688,809,501  $1,634,092,522  
3  $674,460,013  $1,548,588,773  
4  $936,662,225  $1,119,384,610

# 2.3 Observation and Findings

The final version loops through 100 pages, successfully scraping 10,000 movies in total.

The urllib.request.Request method with a User-Agent header prevented 403 errors by imitating a real browser.

The scraper correctly located the table on each page, filtered only valid rows (len(cols)==6)), and stored them in a list.

The addition of time.sleep(1.5) ensured polite scraping and reduced the risk of being blocked.

The resulting DataFrame contained six clean columns: Rank, Year, Title, Worldwide, Domestic, and International.

In [ ]:
# Step 2.4: Run and Review

df.shape




(10000, 6)

In [ ]:
df.isna().sum()


Rank             0
Year             0
Title            0
Worldwide        0
Domestic         0
International    0
dtype: int64

# Step 2.4: Cleaning the Dataset

In [ ]:
import numpy as np
numbers_df = df.copy()

# Step 2.41 Check structure of the dataset
print("Initial shape:", numbers_df.shape)
print(numbers_df.head())

Initial shape: (10000, 6)
  Rank  Year                                 Title       Worldwide  \
0    1  2009                                Avatar  $2,923,706,026   
1    2  2019                     Avengers: Endgame  $2,717,503,922   
2    3  2022              Avatar: The Way of Water  $2,322,902,023   
3    4  1997                               Titanic  $2,223,048,786   
4    5  2015  Star Wars Ep. VII: The Force Awakens  $2,056,046,835   

       Domestic   International  
0  $785,221,649  $2,138,484,377  
1  $858,373,000  $1,859,130,922  
2  $688,809,501  $1,634,092,522  
3  $674,460,013  $1,548,588,773  
4  $936,662,225  $1,119,384,610  




--- I can see here that we have commas and dollar signs which need to be removed




In [ ]:
# Step 2.42: Standarise column names
numbers_df.columns = numbers_df.columns.str.strip().str.title()
print("\nColumn names:", numbers_df.columns)


Column names: Index(['Rank', 'Year', 'Title', 'Worldwide', 'Domestic', 'International'], dtype='object')


In [ ]:
# Step 2.43: Cleaning Financial Columns
money_columns = ["Worldwide", "Domestic", "International"]

for col in money_columns:
    # Convert values to string first (avoids .str error)
    numbers_df[col] = numbers_df[col].astype(str)

    # Remove "$" and "," characters
    numbers_df[col] = numbers_df[col].str.replace("$", "", regex=False)
    numbers_df[col] = numbers_df[col].str.replace(",", "", regex=False)

    # Replace any empty strings or "nan" text with proper NaN values
    numbers_df[col] = numbers_df[col].replace(["", "nan"], np.nan)

    # Convert the column to numeric (float)
    numbers_df[col] = pd.to_numeric(numbers_df[col], errors="coerce")
print("\n After cleaning:")
print(numbers_df.head())


 After cleaning:
  Rank  Year                                 Title   Worldwide     Domestic  \
0    1  2009                                Avatar  2923706026  785221649.0   
1    2  2019                     Avengers: Endgame  2717503922  858373000.0   
2    3  2022              Avatar: The Way of Water  2322902023  688809501.0   
3    4  1997                               Titanic  2223048786  674460013.0   
4    5  2015  Star Wars Ep. VII: The Force Awakens  2056046835  936662225.0   

   International  
0     2138484377  
1     1859130922  
2     1634092522  
3     1548588773  
4     1119384610  


This shows that I have successfully  

*   Removed Currency symbols
*   Removed Comas


*   Replaced empty strings with NAN
*   Numeric Type



In [ ]:
# Step 2.44: Handling Missing Values
print("\nMissing values per column:")
print(numbers_df.isnull().sum())


Missing values per column:
Rank               0
Year               0
Title              0
Worldwide          0
Domestic         100
International      0
dtype: int64


In [ ]:
# 2.51: Check how many rows contain missing values
print("Missing values per column before drop:")
print(numbers_df.isnull().sum())

# 2.52: Remove rows with any missing values
numbers_df = numbers_df.dropna()

# 2.53: Confirm the new dataset shape
print("\n✅ After dropping missing values:")
print("New shape:", numbers_df.shape)

# 2.54: Verify that there are no remaining missing values
print("\nMissing values per column after drop:")
print(numbers_df.isnull().sum())


Missing values per column before drop:
Rank               0
Year               0
Title              0
Worldwide          0
Domestic         100
International      0
dtype: int64

✅ After dropping missing values:
New shape: (9900, 6)

Missing values per column after drop:
Rank             0
Year             0
Title            0
Worldwide        0
Domestic         0
International    0
dtype: int64


In [ ]:
numbers_df.to_csv("clean_boxoffice_numbers.csv")

In [ ]:
# 2.63: Quick verification – reload and preview
check_df = pd.read_csv("clean_boxoffice_numbers.csv")
print("\nPreview of saved file:")
print(check_df.head())

print("\nFile shape:", check_df.shape)


Preview of saved file:
   Unnamed: 0  Rank  Year                                 Title   Worldwide  \
0           0     1  2009                                Avatar  2923706026   
1           1     2  2019                     Avengers: Endgame  2717503922   
2           2     3  2022              Avatar: The Way of Water  2322902023   
3           3     4  1997                               Titanic  2223048786   
4           4     5  2015  Star Wars Ep. VII: The Force Awakens  2056046835   

      Domestic  International  
0  785221649.0     2138484377  
1  858373000.0     1859130922  
2  688809501.0     1634092522  
3  674460013.0     1548588773  
4  936662225.0     1119384610  

File shape: (9900, 7)


# Step 3: Merging